src/Path.hs

```diff
@@ -259,8 +265,13 @@ mkRelFile s =
 -- All directories have a trailing slash, so if you want no trailing
 -- slash, you can use 'System.FilePath.dropTrailingPathSeparator' from
 -- the filepath package.
-toFilePath :: Path b t -> FilePath
-toFilePath (Path l) = l
+toFilePath :: (Typeable b, Typeable t) => Path b t -> FilePath
+toFilePath l =
+  case l of
+    Path [] -> case cast l of
+      Just (_ :: Path Rel Dir) -> curDirNormalizedFP
+      Nothing -> error "Bug: unexpected empty filepath value"
```

Comment:
Can be avoided. The dynamic type check only lets us catch the bug case which is just a defensive programming. But I am curious to know the reasons for avoiding typeable or dynamic type checks?

BTW, why don't we just ban `toFilePath` and use only the APIs with restricted types?